cosmos/x/evm/plugins/state/genesis.go

```diff
@@ -21,38 +21,69 @@
 package state
 
 import (
+	"fmt"
 	"math/big"
+	"sort"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
 
-	"github.com/ethereum/go-ethereum/core"
-
 	"pkg.berachain.dev/polaris/eth/common"
+	"pkg.berachain.dev/polaris/eth/core"
 )
 
 // InitGenesis takes in a pointer to a genesis state object and populates the KV store.
-func (p *plugin) InitGenesis(ctx sdk.Context, ethGen *core.Genesis) {
+func (p *plugin) InitGenesis(ctx sdk.Context, ethGen *core.Genesis) error {
 	p.Reset(ctx)
 
-	// Iterate over the genesis accounts and set the balances.
-	for address, account := range ethGen.Alloc {
-		// TODO: technically wrong since its overriding / hacking the auth keeper and
-		// we are using the nonce from the account keeper as well.
-		p.CreateAccount(address)
-		p.SetBalance(address, account.Balance)
+	// Sort the addresses so that they are in a consistent order.
+	sortedAddresses := make(core.SortableAddresses, 0, len(ethGen.Alloc))
+	for address := range ethGen.Alloc {
+		sortedAddresses = append(sortedAddresses, address)
+	}
+	sort.Sort(sortedAddresses)
+
+	// Iterate over the sorted genesis accounts and set nonces, balances, codes, and storage.
+	for _, address := range sortedAddresses {
+		account := ethGen.Alloc[address]
+
+		// initialize the account on the auth keeper
+		// NOTE: The auth module's init genesis runs before the evm module's init genesis.
+		if p.Exist(address) {
+			// if the account exists on the auth keeper, ensure the nonce is the same.
+			if p.GetNonce(address) != account.Nonce {
+				return fmt.Errorf(
+					"account nonce mismatch for (%s) between auth (%d) and evm (%d) genesis state",
+					address.Hex(), p.GetNonce(address), account.Nonce,
+				)
+			}
+		} else if account.Nonce != 0 {
+			p.SetNonce(address, account.Nonce)
+		}
+
+		// initialize the account data on the state plugin
+		if account.Balance != nil {
+			p.SetBalance(address, account.Balance)
+		}
 		if account.Code != nil {
+			if ethGen.Config.IsEIP155(big.NewInt(0)) && account.Nonce == 0 {
```

Comment:
in my pr this will be killed so maybe handle it another way pre-emptively

Reply:
sure moving to validate genesis